server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -120,25 +120,29 @@ private SearchHits buildSearchHits(SearchContext context, Profiler profiler) {
         List<FetchSubPhaseProcessor> processors = getProcessors(context.shardTarget(), fetchContext, profiler);
         NestedDocuments nestedDocuments = context.getSearchExecutionContext().getNestedDocuments();
 
-        int currentReaderIndex = -1;
-        LeafReaderContext currentReaderContext = null;
+        List<LeafReaderContext> leaves = context.searcher().getIndexReader().leaves();
         LeafNestedDocuments leafNestedDocuments = null;
         CheckedBiConsumer<Integer, FieldsVisitor, IOException> fieldReader = null;
         boolean hasSequentialDocs = hasSequentialDocs(docs);
         SourceLoader.Leaf leafSourceLoader = null;
+        int leafIndex = -1;
+        LeafReaderContext leafReaderContext = null;
+        int endReaderIdx = -1;
```

Comment:
Yeah this still isn't pretty but FetchPhase as a whole is ugly already so I don't think this is making it actively worse. One day I'll wrap this up into a PerSegmentFetchPhase or something.

Reply:
I really like the `LeafBlahBlah` model instead of the `setNextLeaf` model. Not that I'm an expert, but I think the code is generally more readable.